### ISSUE TITLE
is it dead?

### ISSUE THREAD

**Issue by aolko** [btripoloni/RubyCore#3]
subj

**Comment by dklima** [comment]
I am pretty sure, yes, it is 😢

**Comment by aolko** [comment]
Anyone gonna fork & maintain it? (Not a java programmer)

**Comment by ghost** [comment]
https://github.com/general3214/RubyCore

**Comment by nicholasrobertm** [comment]
FYI since the above fork was dead I've forked this https://github.com/nicholasrobertm/RubyCore

I plan to update this to 1.16.5 - I've got it working already with 1.12.2.

**Comment by aolko** [comment]
> FYI since the above fork was dead I've forked this https://github.com/nicholasrobertm/RubyCore
> 
> I plan to update this to 1.16.5 - I've got it working already with 1.12.2.

please fix the readme while you're at it

**Comment by btripoloni** [comment]
Hi, sorry for never reply.

Yes the project is "dead" for some reasons.

1. I don't have time to mantain the project.
2. Minecraft changes just kill the project, everytime time that mojang make any changes I have to change a huge part of my code.
3. Forge is too much complex, apis like Fabric can make the same with less work.
4. The api that I wrote for this project is not good, I have to rewrite all again (the api works good but I see better ways to do that)
5. If i'm going to rewrite all the project I will do this with Fabric.


Also Hytale is comming and mod will be a native thing, so is possble that I focus on Hytale and not on Minecraft anymore, witch depends of how good Hytale will be.
and maybe this make Mojang feel shame and add a true modding api.

Honestly If i'm was going to do this project today, it will be something very diferent.

It's nice that @nicholasrobertm want to continue my work, I can help you with suggestions or explain anything that you want, maybe some code. 
But please don't remove all my credits from the project like you are doing, I spended months doing this while other "experienced" devs saying that it's a impossible thing to do.

**Comment by nicholasrobertm** [comment]
Hey there!

No problem at all, when I initially forked it I looked at your  git history and thought the account was dead / not being used anymore or I would have reached out. Apologies for the credit removal, I wrote a ruby script to update everything in the repo as I needed to update the package name for maintaining the fork. I'll get credit put back into the license / readme which inadvertently got overwritten.

Note on forge & fabric.

When you were last working on this I totally understand where you were coming from. Forge release a shit ton of breaking changes between 1.7.10 and 1.12.2 and updating between them wasn't straightforward. I do like Fabric and some of the mods being made for it, however have concerns of their method of integrating directly with Minecraft. I've tested a ton of Fabric mods together and there is a lot of incompatibility out there due to how fabric works at a core level. I don't see forge going anywhere anytime soon so intend to push ahead, but think fabric support would be a good idea. My general philosophy is to try to make this easier anyway from other work I'm doing.

TLDR:
 - Would love help / to chat about this anytime. My Discord is primerib#3919 - shoot me a friend request on there and we can talk more in dms! 
 - Have a list of current goals & issues I'm facing below.
 - Apologies for the large wall of text 😅 
 - I intend to contribute time to this project each week. I love Ruby and want to see more use of it and since I run a few modded minecraft servers and I'm in the devops world, this seemed like a great way to contribute to the community.

Current goals / issues
- Finding a way to avoid including jruby-complete.jar and the rubycore java code & loader with every mod being built. I'm testing changes for this right now and have gotten it working minus the 'boilerplate' problem with events below. It makes more sense for this to work like a 'dependency' for other ruby mods.

- Be able to have the 'core' be as little code as possible. This combined with fixing the 'bolierplate' section should allow for maintain the library across any minecraft versions, and move the 'updating with minecraft' problem to only be required by the people making mods with it (just like normal). This would also leave room to support things like fabric / spigot at the same time.

- Figuring out how a rubymod is 'bundled' together (to include the .rb files, resources, etc) - Not sure if a zip, or jar file would work better. I'm thinking a jar with the ruby code, resources and a mcmod.info file in it will work best for making forge 'see' it as a mod.

- Moving the 'quickclasses' to a different repo / split on their own. This is where I see the most pain with multiple potential forge / fabric / minecraft versions because essentially these would need maintained like any other forge mod. The point of these I'd think is to wrap the low-level minecraft things (like creating the json files for a block) in a high level easy to use class. This could even be community maintained by everyone making ruby mods, but I see it as separate from the core plugin (to load ruby code from java at all).

- Getting rid of boilerplate 'connection ' code (e.g. the need for rubycore to contain mcp bindings), I have solved part of this (can call objects like Block directly in ruby with no inbetween code) but for the new forge 'events' I've not found a way to do this. Apparently you can use `java_annotation` for this like below:

```
java_annotation 'SubscribeEvent'
def register_renders(event)
  call_method(:register_renders, event)
end
```

But this doesn't seem to currently work. I've asked in the JRuby element.io chat about this and am hoping to find a solution.

Additionally though I've got it working in 1.12.2 when rubycore is loaded a few things pop up in the logs. Despite what the logs say the mod loads fine but it'd be good to figure out what the cause is, at least for the illegal argument exception one.

```
[22:01:29] [Client thread/ERROR] [FML]: There was a problem reading the entry module-info.class in the jar C:\Users\Rob Nicholas\Desktop\MultiMC\instances\RubyTest\.minecraft\mods\rubycore-0.1-all.jar - probably a corrupt zip
net.minecraftforge.fml.common.LoaderException: java.lang.IllegalArgumentException
[22:01:32] [Client thread/WARN] [FML]: MOD HAS DIRECT REFERENCE System.exit() THIS IS NOT ALLOWED REROUTING TO FML!
[22:01:32] [Client thread/WARN] [FML]: Offender: jnr/posix/WindowsPOSIX.childResult(Ljnr/posix/WindowsChildRecord;Z)I
[22:01:32] [Client thread/WARN] [FML]: Use FMLCommonHandler.exitJava instead
[22:01:32] [Client thread/WARN] [FML]: 
```

Expanded explanation:
To explain what I'm trying to do - I'm trying to use some of the new JRuby features to allow getting rid of the 'binding' code all together. The idea is to move away from the current path of:

rubycore.jar -> loader.rb -> mod_example.rb

Where (for example) you need something like:

```
@SubscribeEvent
public static void onRegisterBlocksEvent(@Nonnull final RegistryEvent.Register<Block> event) {
    container.callMethod(core, "register_blocks", event);
}
```
	
In the java definition, and then a subsequent function in loader.rb like this:
```
def register_blocks(event)
  call_method(:register_blocks, event)
end
```

And then also having to define that function in each mod.

From what I could tell from what I've learned of what you did originally it was intended to build any mod you made with this repo as a whole. That's another thing I'm trying to solve which essentially I see looking like this:

rubycore.jar - contains the base java code bundled (via the shadow gradle plugin) with a current version of jruby. 
mod_example.rb/jar - contains everything needed to get started making a mod.

rubycore.jar:
That java binary will only be responsible for calling out to 'loader.rb' which will find any mods in the 'rubymods' directory and load them in but nothing else. Ideally the least amount of 'bindings' as possible. The point here is to treat rubycore as a library that can be used and required for other mods, without having to write all the in between code (e.g. what's in block.rb that maps function names to MCP bindings). Since (from what I understand) modern JRuby lets you use java classes directly from Ruby without bindings, this shouldn't be needed and the mod developer would just need to call the appropriate 'converted' function (see how jruby does this here https://github.com/jruby/jruby/wiki/CallingJavaFromJRuby - essentially if you had setLightLevel(arg) in java - you can call set_light_level(arg) in ruby and it just works)

mod_example.rb/jar:
With the above said, having split rubycore  on it's own separate to the mod, it requires a change in how mods are created. Specifically there are things you need to be bundled with the mod (e.g. the JSON files for blocks, or the resources for your mod). Additionally, while I do wanna throw in the ability to just upload .rb files to a directory on the client or server and run them, it's not very likely to be possible to get forge to show these as 'mods' in game (you'd only see rubycore).

Here's a quick 1.12.2 working example by the way
![image](https://user-images.githubusercontent.com/3313264/120091213-9de8ae80-c0ce-11eb-8ad5-68f0d76fb61b.png)

The rubycore-0.1 mod is actually already split out, the event code is just still needed
![image](https://user-images.githubusercontent.com/3313264/120091231-cf617a00-c0ce-11eb-853d-f40454754358.png)

And the updated example mod:

```
# frozen_string_literal: true

class Example

  def pre_init
    @block = Block.new(:wood)
    @block.set_unlocalized_name('bigboi')
    @block.set_registry_name('test:bigboiblock')
    @block.set_light_level(1.0)

    @itemBlock = ItemBlock.new(@block)
    @itemBlock.set_registry_name('test:bigboiblock')


    @item = Item.new
    @item.set_unlocalized_name('bigboiitem')
    @item.set_registry_name('test:bigboiitem')

    #@tab = QuickTab.new('Tab', @item)

    @item.set_creative_tab(CreativeTabs.tabCombat)
    @block.set_creative_tab(CreativeTabs.tabBlock)

  end

  java_annotation 'SubscribeEvent'
  def register_blocks(event)
		puts "Registering event #{event.to_s}"
        event.getRegistry().register(@block)
		puts "Block registered #{@block.to_s}"
  end

  java_annotation 'SubscribeEvent'
  def register_items(event)
    puts "Registering Item event #{event.to_s}"
    event.getRegistry().register(@itemBlock)
  	event.getRegistry().register(@item)
  end

  java_annotation 'SubscribeEvent'
  def register_renders(event)
    puts "Registering render event #{event.to_s}"
  end


end

add_mod(Example, 'Example', '0.1')
```
